x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyIntegTests.java

```diff
@@ -297,79 +248,117 @@ private void verifyInvalidateResponse(int noOfApiKeys, List<CreateApiKeyResponse
+        ApiKey apiKey = getApiKeyResponseListener.get().getApiKeyInfos()[0];
+        assertThat(apiKey.getId(), is(createdApiKeys.get(1).getId()));
+        assertThat(apiKey.isInvalidated(), is(true));
     }
 
-    public void testExpiredApiKeysDeletedAfter1Week() throws Exception {
-        List<CreateApiKeyResponse> responses = createApiKeys(2, null);
-        Instant created = Instant.now();
+    private void awaitBusySoThatExpiredApiKeysRemoverCanBeTriggered(Client client) throws InterruptedException {
+        final AtomicReference<Long> ref = new AtomicReference<Long>(0L);
+        for (ApiKeyService apiKeyService : internalCluster().getInstances(ApiKeyService.class)) {
+            ref.set((apiKeyService.lastTimeWhenApiKeysRemoverWasTriggered() > ref.get()) ? apiKeyService.lastTimeWhenApiKeysRemoverWasTriggered() : ref.get());
+        }
+        awaitBusy(() -> (client.threadPool().relativeTimeInMillis() - ref.get() < DELETE_INTERVAL_MILLIS), 5, TimeUnit.SECONDS);
+    }
 
+    public void testExpiredApiKeysBehaviorWhenKeysExpired1WeekBeforeAnd1DayBefore() throws Exception {
         Client client = client().filterWithHeader(Collections.singletonMap("Authorization", UsernamePasswordToken
                 .basicAuthHeaderValue(SecuritySettingsSource.TEST_SUPERUSER, SecuritySettingsSourceField.TEST_PASSWORD_SECURE_STRING)));
+        awaitBusySoThatExpiredApiKeysRemoverCanBeTriggered(client);
+
+        int noOfKeys = 4;
+        List<CreateApiKeyResponse> createdApiKeys = createApiKeys(noOfKeys, null);
+        Instant created = Instant.now();
+
         SecurityClient securityClient = new SecurityClient(client);
 
-        AtomicReference<String> docId = new AtomicReference<>();
-        assertBusy(() -> {
-            SearchResponse searchResponse = client.prepareSearch(SecurityIndexManager.SECURITY_INDEX_NAME)
-                    .setSource(SearchSourceBuilder.searchSource().query(QueryBuilders.termQuery("doc_type", "api_key"))).setSize(10)
-                    .setTerminateAfter(10).get();
-            assertThat(searchResponse.getHits().getTotalHits().value, equalTo(2L));
-            docId.set(searchResponse.getHits().getAt(0).getId());
-        });
-        logger.info("searched and found API key with doc id = " + docId.get());
-        assertThat(docId.get(), is(notNullValue()));
-        assertThat(docId.get(), isIn(responses.stream().map(CreateApiKeyResponse::getId).collect(Collectors.toList())));
+        PlainActionFuture<GetApiKeyResponse> getApiKeyResponseListener = new PlainActionFuture<>();
+        securityClient.getApiKey(GetApiKeyRequest.usingRealmName("file"), getApiKeyResponseListener);
+        assertThat(getApiKeyResponseListener.get().getApiKeyInfos().length, is(noOfKeys));
 
+        // Expire the 1st key such that it cannot be deleted by the remover
+        // hack doc to modify the expiration time to a day before
+        Instant dayBefore = created.minus(1L, ChronoUnit.DAYS);
```

Comment:
@bizybot Can we please tie these numbers to some constant in `ExpiredApiKeysRemover`? We could have a public static method in `ExpiredApiKeysRemover` that returns the soonest date before which tokens are considered for removal.

Reply:
> I believe your suggestion to make it configurable is to avoid the doc update hack, is it?

I was actually thinking about readability. Having magic numbers makes me search more in the code, compared to having a constant that I can follow.